docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,323 @@
+**************
+Type Narrowing
+**************
+
+Python programs often contain symbols that take on multiple types within a
+single given scope and that are distinguished by a conditional check at
+runtime. For example, here the variable *name* can be either a ``str`` or
+``None``, and the ``if name is not None`` narrows it down to just ``str``::
+
+    def maybe_greet(name: str | None) -> None:
+        if name is not None:
+            print("Hello, " + name)
+
+This technique is called *type narrowing*.
+To avoid false positives on such code, type checkers understand
+various kinds of conditional checks that are used to narrow types in Python code.
+The exact set of type narrowing constructs that a type checker understands
+is not specified and varies across type checkers. Commonly understood
+patterns include:
+
+* ``if x is not None``
+* ``if x``
+* ``if isinstance(x, SomeType)``
```

Comment:
Would it be useful to have an example of narrowing to a subclass? E.g.

```
def f(x: Base):
  if isinstance(x, SomeImpl): ...
```

Reply:
Since this behavior is unspecified and `isinstance()` behavior does vary across type checkers, I'd prefer not to go into too much detail.